client-common/src/tendermint/rpc_client/async_rpc_client.rs

```diff
@@ -0,0 +1,262 @@
+    where
+        for<'de> T: Deserialize<'de>,
+    {
+        let response_value = self.request(method, params).await?;
+        serde_json::from_value(response_value).with_context(|| {
+            format!(
+                "Unable to deserialize `{}` from JSON-RPC response for params: {:?}",
+                method, params
+            )
+        })
+    }
+
+    /// Makes RPC call in batch and deserializes responses
+    ///
+    /// TODO: Usage of `Vec` can be removed once we execute it in a purely async context
+    pub async fn call_batch<T>(&self, batch_params: &[(&str, Vec<Value>)]) -> Result<Vec<T>>
+    where
+        for<'de> T: Deserialize<'de>,
+    {
+        let response_values = self.request_batch(batch_params).await?;
+        let mut responses = Vec::with_capacity(response_values.len());
+
+        for (i, response_value) in response_values.into_iter().enumerate() {
+            let method = batch_params[i].0;
+            let params = &batch_params[i].1;
+
+            let response = serde_json::from_value(response_value).with_context(|| {
+                format!(
+                    "Unable to deserialize `{}` from JSON-RPC response for params: {:?}",
+                    method, params
+                )
+            })?;
+
+            responses.push(response);
+        }
+
+        Ok(responses)
+    }
+
+    /// Sends a JSON-RPC request and returns `request_id` and `response_channel`
+    async fn send_request(
+        &self,
+        method: &str,
+        params: &[Value],
+    ) -> Result<(String, Receiver<JsonRpcResponse>)> {
+        let id = self.unique_id.fetch_add(1, Ordering::Relaxed).to_string();
+        let message = prepare_message(&id, method, params)?;
+        let (channel_sender, channel_receiver) = channel::<JsonRpcResponse>();
+
+        self.channel_map
+            .lock()
+            .await
+            .insert(id.clone(), channel_sender);
+
+        self.ensure_connected().await?;
+
+        if let Err(err) = self
+            .websocket_writer
+            .lock()
+            .await
+            .send(message)
+            .await
+            .context("Unable to send message to websocket writer")
+        {
+            self.channel_map.lock().await.remove(&id);
+            bail!(err);
+        }
+
+        Ok((id, channel_receiver))
+    }
+
+    /// Receives response from websocket for given id.
+    async fn receive_response(
+        &self,
+        method: &str,
+        params: &[Value],
+        id: &str,
+        receiver: Receiver<JsonRpcResponse>,
+    ) -> Result<Value> {
+        let response = timeout(RESPONSE_TIMEOUT, receiver)
+            .await
+            .context("Tendermint RPC request timed out")?;
+
+        let response = match response.context("Unable to receive message from channel receiver") {
+            Ok(response) => response,
+            Err(err) => {
+                self.channel_map.lock().await.remove(id);
+                bail!(err)
```

Comment:
those existing error chaining helpers in client-* don't work in async?

Reply:
Yes. `client-common::Error` is not `Send + Sync`. We can slowly start moving everything to `anyhow` (in `client-*`) or `thiserror` (in `chain-*`).